kimchi/src/circuits/cairo/runner/program.rs

```diff
@@ -0,0 +1,466 @@
+    /// This function creates an instance of a default CairoVariables struct
+    pub fn new() -> CairoVariables<F> {
+        CairoVariables {
+            dst: None,
+            op0: None,
+            op1: None,
+            res: None,
+            dst_addr: F::zero(),
+            op0_addr: F::zero(),
+            op1_addr: F::zero(),
+            size: F::zero(),
+        }
+    }
+}
+impl<F: PrimeField> Default for CairoVariables<F> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+/// A data structure to store a current step of Cairo computation
+pub struct CairoStep<'a, F: PrimeField> {
+    // current step of computation
+    //step: u64,
+    /// current word of the program
+    pub mem: &'a mut CairoMemory<F>,
+    // comment instr for efficiency
+    /// current pointers
+    curr: CairoPointers<F>,
+    /// (if any) next pointers
+    next: Option<CairoPointers<F>>,
+    /// state auxiliary variables
+    vars: CairoVariables<F>,
+}
+
+/*
+/// Performs the addition of a u64 register with a signed offset
+fn add_off(reg: u64, off: i128) -> u64 {
+    // TODO(@querolita) check helper to avoid casting manually
+    // 1. Since the output address is u64, the offset cannot be larger than this
+    // 2. The absolute value of the offset will fit in u64
+    // 3. If it was negative, subtract. If it was positive, sum.
+    if off.is_negative() {
+        reg - (off.abs() as u64)
+    } else {
+        reg + (off.abs() as u64)
+    }
+}
+*/
+
+impl<'a, F: PrimeField> CairoStep<'a, F> {
+    /// Creates a new Cairo execution step from a step index, a Cairo word, and current pointers
+    pub fn new(mem: &mut CairoMemory<F>, ptrs: CairoPointers<F>) -> CairoStep<F> {
+        CairoStep {
+            //step: idx,
+            mem,
+            curr: ptrs,
+            next: None,
+            vars: CairoVariables::new(),
+        }
+    }
+
+    /// Executes a Cairo step from the current registers
+    pub fn execute(&mut self) {
+        // This order is important in order to allocate the memory in time
+        self.set_op0();
+        self.set_op1();
+        self.set_res();
+        self.set_dst();
+        // If the Option<> thing is not a guarantee for continuation of the program, we may be removing this
+        let next_pc = self.next_pc();
+        let (next_ap, next_fp) = self.next_apfp();
+        self.next = Some(CairoPointers::new(
+            next_pc.unwrap(),
+            next_ap.unwrap(),
+            next_fp.unwrap(),
+        ));
+    }
+
+    /// This function returns the current word instruction being executed
+    pub fn instr(&mut self) -> CairoWord<F> {
+        CairoWord::new(self.mem.read(self.curr.pc).unwrap())
+    }
+
+    /// This function computes the first operand address
+    pub fn set_op0(&mut self) {
+        if self.instr().op0_reg() == OP0_AP {
+            // reads first word from allocated memory
+            self.vars.op0_addr = self.curr.ap + self.instr().off_op0();
```

Comment:
I guess it all depends on what assumptions we are making on the kind of input we would be executing the runner with. The input should be a program that was compiled trustfully, and thus should not generate instructions accessing invalid addresses of memory. Meaning, we should not be executing the Cairo runner on field elements that do not correspond to real bytecode computations. So if this is the case, then I think we don't need to worry about over/underflows. Now, if we wanted this piece of code to be "safe enough", then we should decide what should be a reasonable limit for the memory size for instance. We are already doing this implicitly, as we are casting the address to a `u64`. But I guess this size is already too large and would cause memory overflows in any case. Even `u32` seems too large, but then `u16` may be too small.... Any thoughts on this?